File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/DistanceSensors.java
package org.firstinspires.ftc.teamcode.subsystems;

import com.arcrobotics.ftclib.command.SubsystemBase;
import com.qualcomm.hardware.rev.Rev2mDistanceSensor;
import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.util.ElapsedTime;

import org.firstinspires.ftc.robotcore.external.navigation.DistanceUnit;
import org.firstinspires.ftc.teamcode.util.MB1242;
import org.outoftheboxrobotics.neutrinoi2c.Rev2mDistanceSensor.AsyncRev2MSensor;

public class DistanceSensors extends SubsystemBase {

    //forward sensors
    private final MB1242 forwardSensor;
    private final MB1242 backwardSensor;
    //The TOF Distance sensor on the sides.
    private final AsyncRev2MSensor leftSensor;

    //Debugging rate timer
    private final ElapsedTime cycleTimer = new ElapsedTime(ElapsedTime.Resolution.MILLISECONDS);
    private double cycleTime = 0.0;

    public DistanceSensors(HardwareMap hardwareMap) {

        //Get the sensors from the hardware map
        //TODO: YES I KNOW ITS REVERSED BUT I REFUSE TO SWITCH THE WIRING AROUND NOW
        forwardSensor = hardwareMap.get(MB1242.class, "forwardSensor");
        backwardSensor = hardwareMap.get(MB1242.class, "backwardSensor");
        leftSensor =
                new AsyncRev2MSensor(hardwareMap.get(Rev2mDistanceSensor.class, "leftSensor"));

        leftSensor.setMeasurementIntervalMs(60);

//        disableAll();
        cycleTimer.reset();

    }

    public void pingAll(){
        forwardSensor.ping();
        backwardSensor.ping();
    }

//    public void enableAll(){
//        leftSensor.enable();
//        forwardSensor.enable();
//        backwardSensor.enable();
//    }
//
//    public void disableAll(){
//        leftSensor.disable();
//        forwardSensor.disable();
//        backwardSensor.disable();
//    }

    @Override
    public void periodic() {
        cycleTime = cycleTimer.milliseconds();
        cycleTimer.reset();
    }


    //Get the left sensor range in cm
    public double getLeftRange(DistanceUnit unit) {
        return unit.fromCm(leftSensor.getDistance(DistanceUnit.CM));
    }

    public double getForwardRange(DistanceUnit unit) {
        return unit.fromCm(forwardSensor.getDistance(DistanceUnit.CM));
    }

    public double getBackwardRange(DistanceUnit unit) {
        return unit.fromCm(backwardSensor.getDistance(DistanceUnit.CM));
    }

    /**
     * Returns the cycle time in milliseconds of the distance sensors.
     */
    public double getCycleTime() {
        return cycleTime;
    }

    //Tests the sensor by running a range command and seeing if there's an output, takes at least 100ms
    public boolean test() {
        ElapsedTime timer = new ElapsedTime();
        while (timer.milliseconds() < 80) ;
        return forwardSensor.getDistance(DistanceUnit.CM) > 20 || //20cm is the minimum range, so we test with it
                leftSensor.getDistance(DistanceUnit.CM) > 3 ||
                backwardSensor.getDistance(DistanceUnit.CM) > 20;
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/DistanceSensors.kt
package org.firstinspires.ftc.teamcode.subsystems

import com.arcrobotics.ftclib.command.SubsystemBase
import com.qualcomm.hardware.rev.Rev2mDistanceSensor
import com.qualcomm.robotcore.hardware.HardwareMap
import com.qualcomm.robotcore.util.ElapsedTime
import org.firstinspires.ftc.robotcore.external.Telemetry
import org.firstinspires.ftc.robotcore.external.navigation.DistanceUnit
import org.firstinspires.ftc.teamcode.util.MB1242

/*
When we do SubsystemBase() in kotlin here, it calls the super() constructor.
 */
class DistanceSensors(private val hardwareMap: HardwareMap, private val telemetry: Telemetry? = null): SubsystemBase() {


    //Distance sensors

    //The ultrasonic sensor facing forward
    private val forwardSensor: MB1242

    //The two TOF Distance sensors on the sides.
    private val rightSensor: Rev2mDistanceSensor
    //TODO: Uncomment this (cursor on line and then press Ctrl + /)
    private val leftSensor: Rev2mDistanceSensor


    //Timer object for the delay
    private val delayTimer = ElapsedTime()

    //Delay in ms between pings, recommended 80-100ms for the ultrasonic sensor
    private val readingDelayMs: Int = 80

    //Whether we should be actively reading the range or not
    private var takingRangeReading = false

    //Last read range in cm
    private var forwardRangeReading = 0.0
    private var leftRange = 0.0
    private var rightRange = 0.0

    //Whether this is the first run of the program
    private var firstRun = true

    init {
        //Get the sensors from the hardware map
        forwardSensor = hardwareMap.get(MB1242::class.java, "forwardSensor")
        rightSensor = hardwareMap.get(Rev2mDistanceSensor::class.java, "rightSensor")
        leftSensor = hardwareMap.get(Rev2mDistanceSensor::class.java,"leftSensor")
    }

    //Gets updated constantly throughout auto.
    override fun periodic() {
        //Only take readings if we are supposed to, it's an unnecessary hardware call otherwise
        if(takingRangeReading){
            //Check if its been longer than our delay to properly let the sensor perform
            if(hasDelayExpired() && !firstRun){
                forwardRangeReading = forwardSensor.readRange().toDouble()
                //TODO: Read and store the left and right sensors IN CENTIMETERS (DistanceUnit.CM)
                rightRange = rightSensor.getDistance(DistanceUnit.CM)
                leftRange = leftSensor.getDistance(DistanceUnit.CM)
                forwardSensor.ping()
                delayTimer.reset()
                //If its the first run then run a range command to ensure the next reading has a value
            } else if(firstRun){
                forwardSensor.ping()
                delayTimer.reset()
                firstRun = false
            }
        }

        //TODO: If   you want, add telemetry calls here with the following
        telemetry?.addData("foo", 3);
        //Telemetry is updated later on probably
    }



    fun getRanges(redSide: Boolean): ArrayList<Double> {
        val list = ArrayList<Double>();

        list.add(forwardRangeReading);
        if (redSide) {
            list.add(rightRange);
        }
        else {
            list.add(leftRange);
        }

        return list;
    }





    //TODO: Add get range methods for the left and right sensors
    fun getLeftRange(): Double {
        return leftRange
    }

    fun getRightRange(): Double {
        return rightRange
    }

    /**
     * Read the current range value in cm, is only updated every 80ms so might be a little inaccurate
     */
    fun getForwardRange(): Double {
        return forwardRangeReading
    }

    /**
     * Returns the distance in the specified value
     */
    fun getForwardRange(distanceUnit: DistanceUnit): Double{
        return distanceUnit.fromCm(forwardRangeReading)
    }


    //Start taking range readings
    fun startReading(){
        takingRangeReading = true
    }

    //Stop taking range readings
    fun stopReading(){
        takingRangeReading = false
    }


    //Tells if the delay timer has expired or not
    fun hasDelayExpired(): Boolean {
        return delayTimer.milliseconds() >= readingDelayMs
    }


    //Tests the sensor by running a range command and seeing if there's an output, takes at least 100ms
    fun test(): Boolean {
        forwardSensor.ping()
        val timer = ElapsedTime()
        while(timer.milliseconds() < 100);
        return forwardSensor.readRange().toInt() != 20 //20cm is the minimum range, so we test with it
    }

}